signer/signer_test.go

```diff
@@ -662,3 +664,262 @@ func TestMakeKeyGCPRSA(t *testing.T) {
+
+	mockFactory := mockedPKCS11ContextFactory(mockCtx)
+	crypto11.Configure(&crypto11.PKCS11Config{}, mockFactory)
+	defer crypto11.Close()
+
+	for _, testcase := range unlimitedBytesReaderPassingTestCases {
+		var expectedBytes []byte
+		for _, bytes := range testcase.expectedPkcs11RandReaderBytes {
+			expectedBytes = append(expectedBytes, bytes...)
+			mockCtx.EXPECT().GenerateRandom(session, len(bytes)).Return(bytes, nil).Times(1)
+		}
+		result := make([]byte, testcase.requestedBytes)
+		n, err := ubrr.Read(result)
+		if err != nil {
+			t.Fatalf("unlimitedBytesRandReader.Read failed: %v", err)
+		}
+		if n != testcase.requestedBytes {
+			t.Fatalf("failed to read %d bytes, read %d instead", testcase.requestedBytes, n)
+		}
+		if !bytes.Equal(result, expectedBytes) {
+			t.Fatalf("result is not the expected bytes. got: %v, want: %v", result, expectedBytes)
+		}
+	}
+	// these ones are called as part of Close(), not as part of our actual testing
+	mockCtx.EXPECT().CloseSession(session).Return(nil).Times(expectedSessions)
+	mockCtx.EXPECT().CloseAllSessions(slot).Return(nil).Times(1)
+	mockCtx.EXPECT().Finalize().Return(nil).Times(1)
+	mockCtx.EXPECT().Destroy().Times(1)
+}
+
+// inputs and outputs to/from the under the hood GenerateRandom calls
+type grExpectation struct {
+	requestedBytes int
+	returnedBytes  []byte
+}
+
+var unlimitedBytesReaderFailingTestCases = []struct {
+	// number of random bytes to request
+	requestedBytes int
+	// the actual bytes we expect back. this will always be the same
+	// length as requestedBytes, and will be filled with the actual bytes
+	// returned for success calls to PKCS11RandReader.Read, followed by zeros
+	// this is true even when GenerateRandom returns some (but not all of the
+	// requested) random bytes, because PKCS11RandReader.Read throws the
+	// returned bytes from GenerateRandom on any error
+	expectedBytes              []byte
+	generateRandomExpectations []grExpectation
+}{
+	{
+		requestedBytes: 5,
+		expectedBytes:  bytes.Repeat([]byte{0}, 5),
+		generateRandomExpectations: []grExpectation{
+			{
+				requestedBytes: 5,
+				returnedBytes:  []byte{},
+			},
+		},
+	},
+	{
+		requestedBytes: 1024,
+		expectedBytes:  bytes.Repeat([]byte{0}, 1024),
+		generateRandomExpectations: []grExpectation{
+			{
+				requestedBytes: 1024,
+				returnedBytes:  bytes.Repeat([]byte("A"), 5),
+			},
+		},
+	},
+	{
+		requestedBytes: 2048,
+		expectedBytes: append(
+			bytes.Repeat([]byte("A"), 1024),
+			bytes.Repeat([]byte{0}, 1024)...,
+		),
+		generateRandomExpectations: []grExpectation{
+			{
+				requestedBytes: 1024,
+				returnedBytes:  bytes.Repeat([]byte("A"), 1024),
+			},
+			{
+				requestedBytes: 1024,
+				returnedBytes:  bytes.Repeat([]byte("B"), 72),
+			},
+		},
+	},
+	{
+		requestedBytes: 5000,
+		expectedBytes: append(
+			append(
+				bytes.Repeat([]byte("A"), 1024),
+				bytes.Repeat([]byte("B"), 1024)...,
+			),
+			bytes.Repeat([]byte{0}, 2952)...,
+		),
```

Comment:
There may be a better way to generate these inline, but I couldn't find it.

Reply:
yeah, the only think i could think of was a "pre-processing" for-loop that turned sizes into slices and did byte math with the iterator (`byte('a')+i`) between the test cases and the actual uses. But that has the downsize of not being explicit in what folks should see. I like this better!